controllers/provider-aws/pkg/controlplane/actuator.go

```diff
@@ -0,0 +1,343 @@
+				{
+					"name": "AWS_ACCESS_KEY_ID",
+					"valueFrom": map[string]interface{}{
+						"secretKeyRef": map[string]interface{}{
+							"key":  AccessKeyID,
+							"name": common.CloudProviderSecretName,
+						},
+					},
+				},
+				{
+					"name": "AWS_SECRET_ACCESS_KEY",
+					"valueFrom": map[string]interface{}{
+						"secretKeyRef": map[string]interface{}{
+							"key":  SecretAccessKey,
+							"name": common.CloudProviderSecretName,
+						},
+					},
+				},
+			},
+			"resources": map[string]interface{}{
+				"limits": map[string]interface{}{
+					"cpu":    "500m",
+					"memory": "512Mi",
+				},
+			},
+		}, nil
+	},
+	Objects: []*chart.Object{
+		{Type: &corev1.Service{}, Name: "cloud-controller-manager"},
+		{Type: &appsv1.Deployment{}, Name: "cloud-controller-manager"},
+	},
+}
+
+// NewActuator creates a new Actuator that acts upon and updates the status of ControlPlane resources.
+func NewActuator() controlplane.Actuator {
+	return &actuator{
+		logger: log.Log.WithName("controlplane-controller"),
+	}
+}
+
+// actuator is an Actuator that acts upon and updates the status of ControlPlane resources.
+type actuator struct {
+	scheme            *runtime.Scheme
+	decoder           runtime.Decoder
+	config            *rest.Config
+	clientset         kubernetes.Interface
+	gardenerClientset gardenerkubernetes.Interface
+	chartApplier      gardenerkubernetes.ChartApplier
+	client            client.Client
+	logger            logr.Logger
+}
+
+// InjectScheme injects the given scheme into the actuator.
+func (a *actuator) InjectScheme(scheme *runtime.Scheme) error {
+	a.scheme = scheme
+	a.decoder = serializer.NewCodecFactory(a.scheme).UniversalDecoder()
+	return nil
+}
+
+// InjectConfig injects the given config into the actuator.
+func (a *actuator) InjectConfig(config *rest.Config) error {
+	a.config = config
+
+	// Create clientset
+	var err error
+	a.clientset, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "could not create Kubernetes client")
+	}
+
+	// Create Gardener clientset
+	a.gardenerClientset, err = gardenerkubernetes.NewForConfig(config, client.Options{})
+	if err != nil {
+		return errors.Wrap(err, "could not create Gardener client")
+	}
+
+	// Create chart applier
+	a.chartApplier, err = gardenerkubernetes.NewChartApplierForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "could not create chart applier")
+	}
+
+	return nil
+}
+
+// InjectClient injects the given client into the actuator.
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+// Reconcile reconciles the given controlplane and cluster, creating or updating the additional Shoot
+// control plane components as needed.
+func (a *actuator) Reconcile(
+	ctx context.Context,
+	cp *extensionsv1alpha1.ControlPlane,
+	cluster *extensionscontroller.Cluster,
+) error {
+	// Decode providerConfig
+	cpConfig := &apisaws.ControlPlaneConfig{}
```

Comment:
This method is very big and complex - please split it into various methods.

Reply:
"Very big and complex" is very subjective in my opinion. Could you be somewhat more specific regarding what should be extracted as functions, and why?
I personally am not a big fan of extracting random pieces of code just to make methods short. A longer methods that has no branch complexity is clearer and easier to understand than the same method cut into scattered pieces that have to be pieced together again by the reader each time.

Reply:
One piece of code that could be extracted would be determining the `subnetID` and `zone` from the infrastructure status. Another one to consider would be computing the checksums, but I am somewhat reluctant about this one, I personally prefer to see the entire flow in one place since the steps depend on each other.

Reply:
Extracted a few methods out of it now, basically everything that I think would make sense to extract. Hopefully it's less "very big and complex" now.